lib/src/highlighter.dart

```diff
@@ -442,9 +444,42 @@ class Highlighter {
       ..write('^');
   }
 
-  /// Writes a space followed by [label] if [label] isn't `null`.
-  void _writeLabel(String? label) {
-    if (label != null) _buffer.write(' $label');
+  /// Writes [highlight]'s label.
+  ///
+  /// The `_buffer` is assumed to be written to the point where the first line
+  /// of `highlight.label` can be written after a space, but this takes care of
+  /// writing indentation and highlight columns for later lines.
+  ///
+  /// The [underlineLength] is the length of the line written between the
+  /// highlights and the beginning of the first label.
+  void _writeLabel(_Highlight highlight, List<_Highlight?> highlightsByColumn,
+      int underlineLength) {
+    final label = highlight.label;
+    if (label == null) {
+      _buffer.writeln();
+      return;
+    }
+
+    final lines = label.split('\n');
+    final color = highlight.isPrimary ? _primaryColor : _secondaryColor;
+    _colorize(() => _buffer.write(' ${lines.first}'), color: color);
+    _buffer.writeln();
+
+    for (var text in lines.skip(1)) {
+      _writeSidebar();
+      _buffer.write(' ');
+      for (var columnHighlight in highlightsByColumn) {
```

Comment:
Writing a line is always going to be O(columns), because a line is inherently `columns` characters long, so I don't think this changes the complexity class. I don't think this is going to be a problem, since this code is primarily for human reading and thus unlikely to be performance-critical and it's extremely unlikely that `highlightsByColumn` is more than 3 elements long anyway.